Custom SQL Models (#426)
rudderlabs/rudderstack-docs: connections/models.mdx

---
# slug: "/docs/rudderstack-cloud/models"
title: "Models"
description: >-
  Detailed guide on custom SQL models and how to create, update and fetch their details in RudderStack dashboard.
---

# Models

The **Models** feature lets you create models by defining custom SQL queries. These queries can then be executed over your warehouse to send the resultant data to the specified destinations. You can create as many models and reuse them with their corresponding Warehouse Actions sources.

<div class="infoBlock">

> Currently, the <strong>Models</strong> feature is supported for <a href="https://www.rudderstack.com/docs/warehouse-actions/amazon-redshift/">Amazon Redshift</a>, <a href="https://www.rudderstack.com/docs/warehouse-actions/google-bigquery/">Google BigQuery</a>, <a href="https://www.rudderstack.com/docs/warehouse-actions/postgresql/">PostgreSQL</a>, and <a href="https://www.rudderstack.com/docs/warehouse-actions/snowflake/">Snowflake</a> Warehouse Actions sources only.

Reviewer: ```suggestion
Currently, the <strong>Models</strong> feature is supported for the following Warehouse Actions sources:
<ul>
<li><a href="https://www.rudderstack.com/docs/warehouse-actions/amazon-redshift/">Amazon Redshift</a></li> 
<li><a href="https://www.rudderstack.com/docs/warehouse-actions/google-bigquery/">Google BigQuery</a></li>
<li><a href="https://www.rudderstack.com/docs/warehouse-actions/postgresql/">PostgreSQL</a> </li>
<li><a href="https://www.rudderstack.com/docs/warehouse-actions/snowflake/">Snowflake</a></li>
</ul>
```

Reviewer: Presenting this as a list will be better IMO.

</div>

> The RudderStack dashboard provides **Models** option in the left navigation bar that allows you to view all the existing models:

Reviewer: ```suggestion
You can find the **Models** option in the left navigation bar of the RudderStack dashboard that lets you view all the existing models, as shown:
```

<img src="../assets/warehouse-actions-sources/models-view-dashboard.png" alt="RudderStack Models" />

> ## Create model

Reviewer: ```suggestion
## Creating a new model
```

Follow these steps to create a model:

1. Log in to the **RudderStack dashboard**.

2. Click on **Models** in the left navigation bar and select **New model**.

<img src="../assets/warehouse-actions-sources/models-dashboard.png" alt="RudderStack Models"/>

> 3. Select the type of Warehouse Action for the model and click **Next**.

Reviewer: ```suggestion
3. Select the Warehouse Actions source for the model and click **Next**.
```

<img src="../assets/warehouse-actions-sources/select-source-model.png" alt="Add Model"/>

4. Enter a name for the model.

> 5. Select from the existing credentials or create a new warehouse account.

Reviewer: ```suggestion
5. Select your existing warehouse credentials or click on **Create Credentials from Scratch** and enter the warehouse-specific details.
```

 
<div class="infoBlock">

> The warehouse account must have the necessary permissions to be integrated with RudderStack. For more information on these permissions, refer to the <strong>Granting Permissions</strong> section of the relevant <a href="https://www.rudderstack.com/docs/warehouse-actions/">Warehouse Actions</a> source.

Reviewer: ```suggestion
Your warehouse account must have the necessary user permissions for RudderStack to use it to create the models. For more information on these user permissions, refer to the <strong>Granting Permissions</strong> section of the relevant <a href="https://www.rudderstack.com/docs/warehouse-actions/">Warehouse Actions</a> source.
```

  
</div>

> 

Reviewer: Should we add a note about the validations here?

<img src="../assets/warehouse-actions-sources/model-settings.png" alt="Model settings" />

> 6. Enter the **Description** and custom SQL **Query**. Click on **Run Query** to **Preview** the result.

Reviewer: We already talk about the **Preview** option in the note below so we don't need to repeat it here. Thoughts?

<img src="../assets/warehouse-actions-sources/sample-sql-query.png" alt="Sample SQL query" />

<div class="infoBlock">

The <strong>Preview</strong> section displays the preview of the 50 resultant rows in a paginated format.
</div>

> 7. Click **Next** to create the model succesfully.

Reviewer: ```suggestion
8. Click on **Next** to create the model successfully.
```

<div class="infoBlock">
Once a model is created, you can use it to send the data from a warehouse source to a specified destination. For more information and detailed steps, refer to <a href="https://www.rudderstack.com/docs/warehouse-actions/setting-connections-using-models/">Setting Connections using Models</a>.
</div>

> ## Update model

Reviewer: ```suggestion
## Updating an existing model
```

Follow these steps to update an existing model:

1. Click the model you want to update and select the **Configuration** tab.

> 2. Click on **Edit Configuration** and make the required changes in  **Description** and/or **Query**.

Reviewer: ```suggestion
2. Click on **Edit Configuration** and make the required query changes under **Query**.
```

> 3. If you have edited the **Query**, click on **Run Query** to see if the new query generates the desired result.

Reviewer: ```suggestion
3. Upon making the changes, click on **Run Query** to verify if you get the desired result.
```

4. Finally, click on **Finish** to save your changes.

<img src="../assets/warehouse-actions-sources/edit-model-configuration.png" alt="Edit existing Model configuration" />

<div class="infoBlock">
  

> If you update a model's <strong>Query</strong>, you also need to check or update the JSON and <a href="https://www.rudderstack.com/docs/warehouse-actions/visual-data-mapper/">Visual Data Mapper</a> mapping to trigger the full sync and validate a sync by manually triggering it.

Reviewer 2: May be, I wasnt clear on this part.
JSON Mapper and Visual Data Mapper are 2 separate mappers.
should be `or` in between.

Author: understood and updated

> If you update a model's <strong>Query</strong>, you also need to check/update the JSON Mapper or <a href="https://www.rudderstack.com/docs/warehouse-actions/visual-data-mapper/">Visual Data Mapper</a> mapping to trigger the full sync and validate a sync by manually triggering it.

Reviewer: ```suggestion
If you update a model's query, you also need to check and update the relevant JSON mapping or <a href="https://www.rudderstack.com/docs/warehouse-actions/visual-data-mapper/">Visual Data Mapping</a>. To validate if the query is running fine, you can manually trigger a full sync.
```

> </div>

Reviewer: have some doubts about this - pending confirmation from @ashishRudder

## Model details

To get details about an existing model, you can click on it to see the options listed below:

- **Rename**: Lets you rename the model by clicking on the edit icon next to the model's name.

<img src="../assets/warehouse-actions-sources/model-rename.png" alt="Rename model" />

- **Sources**: Displays the sources associated with the model. To know more on how to use a model with a source, refer to [Setting Connections using Models](https://www.rudderstack.com/docs/warehouse-actions/setting-connections-using-models/).

- **Configuration**: Displays and lets you edit the SQL **Query** and **Description**.

> - **Settings**: This option provides additional model-related details, such as:

Reviewer: ```suggestion
- **Settings**: This option contains the following two settings:
```

---
# slug: "/docs/rudderstack-cloud/models"
title: "Models"
description: >-
  Detailed guide on custom SQL models and how to create, update and fetch their details in RudderStack dashboard.
---

# Models

> The **Models** feature lets you create models by defining custom SQL queries. These queries can then be executed over the Warehouse sources schema to send the resultant data to the specified destinations. You can create as many models and reuse them with their corresponding Warehouse Actions sources.

Reviewer 2: `Warehouse sources schema` is not valid since the tables/views used in the query can be in different schemas.
it should be the term `over your warehouse`. it adds more clarity.

<div class="infoBlock">

  Currently, the <strong>Models</strong> feature is supported for <a href="https://www.rudderstack.com/docs/warehouse-actions/amazon-redshift/">Amazon Redshift</a>, <a href="https://www.rudderstack.com/docs/warehouse-actions/google-bigquery/">Google BigQuery</a>, <a href="https://www.rudderstack.com/docs/warehouse-actions/postgresql/">PostgreSQL</a>, and <a href="https://www.rudderstack.com/docs/warehouse-actions/snowflake/">Snowflake</a> Warehouse Actions sources only.
</div>

The RudderStack dashboard provides **Models** option in the left navigation bar that allows you to view all the existing models:

<img src="../assets/warehouse-actions-sources/models-view-dashboard.png" alt="RudderStack Models" />

## Create model

Follow these steps to create a model:

1. Log in to the **RudderStack dashboard**.

2. Click on **Models** in the left navigation bar and select **New model**.

<img src="../assets/warehouse-actions-sources/models-dashboard.png" alt="RudderStack Models"/>

> 3. Select a source for the model and click **Next**.

Reviewer 2: `a souce` is a different object from Rudderstack's perspective.
This can be changed to ` the type of Warehouse action`.

<img src="../assets/warehouse-actions-sources/select-source-model.png" alt="Add Model"/>

4. Configure the model with relevant source settings and click **Next**. For more information on these settings, refer [Warehouse Actions Sources](https://www.rudderstack.com/docs/warehouse-actions/).

<img src="../assets/warehouse-actions-sources/model-settings.png" alt="Model settings" />

5. Enter the **Description** and custom SQL **Query**. Click on **Run Query** to **Preview** the result.

<img src="../assets/warehouse-actions-sources/sample-sql-query.png" alt="Sample SQL query" />

<div class="infoBlock">

> The <strong>Preview</strong> section displays the preview of the first 50 resultant rows in a paginated format.

Reviewer 3: minor: `the preview of the first 50 resultant rows` -> is this technically right? We do limit 50 I guess. Does this mean first 50 rows for all WHs? We can remove the term `first` if its not technically correct. @vyeshwanth

Reviewer 4: Yes,  we can remove `first`.

</div>

6. Click **Next** to configure the model succesfully.

> ## Update model

Reviewer 3: Should we highlight, after changing model query, check/update VDM mapping to trigger full sync and also validate a sync by manually triggering?

Follow these steps to update an existing model:

1. Click the model you want to update and select the **Configuration** tab.
2. Click on **Edit Configuration** and make the required changes in  **Description** and/or **Query**.
3. If you have edited the **Query**, click on **Run Query** to see if the new query generates the desired result.
4. Finally, click on **Finish** to save your changes.

<img src="../assets/warehouse-actions-sources/edit-model-configuration.png" alt="Edit existing Model configuration" />

## Model details

To get details about an existing model, you can click on it to see the options listed below:

- **Rename**: Lets you rename the model by clicking on the edit icon next to the model's name.

> <img src="../assets/warehouse-actions-sources/model-rename.png" alt="Rename source" />

Reviewer 2: not sure if its required but `alt` should be `Rename model`

- **Sources**: Displays the sources associated with the model. To know more on how to use a model with a source, refer to [Custom SQL Models](https://www.rudderstack.com/docs/warehouse-actions/custom-sql-models/).

- **Configuration**: Displays and lets you edit the SQL **Query** and **Description**.

- **Settings**: This option provides additional model-related details, such as:  

  - **Credentials**: Lets you edit the credentials used for the source.

>   - **Permanently delete the model**: Lets you delete the model by clicking on the **Delete** button. The **Delete** button is only enabled when no source is connected to the model.

Reviewer 2: `is connected to the model.` -> `is using the model`.

</div>

The RudderStack dashboard provides **Models** option in the left navigation bar that allows you to view all the existing models:

<img src="../assets/warehouse-actions-sources/models-view-dashboard.png" alt="RudderStack Models" />

## Create model

Follow these steps to create a model:

1. Log in to the **RudderStack dashboard**.

2. Click on **Models** in the left navigation bar and select **New model**.

<img src="../assets/warehouse-actions-sources/models-dashboard.png" alt="RudderStack Models"/>

3. Select the type of Warehouse Action for the model and click **Next**.

<img src="../assets/warehouse-actions-sources/select-source-model.png" alt="Add Model"/>

4. Enter a name for the model.
5. Select from the existing credentials or create a new warehouse account.
 
<div class="infoBlock">

> The warehouse account must have the necessary permissions to be integrated with RudderStack. For more information on these permissions, refer to the <a href="https://www.rudderstack.com/docs/warehouse-actions/">Granting Permissions</a> section of the relevant Warehouse Actions source.

Reviewer 2: 1. Instead of having hyperlink to `Granting Permissions`, we can have it in bold.
2. And add hyperlink over `Warehouse Actions` since the hyper link wont take u to Granting Permission section.

  
</div>

<img src="../assets/warehouse-actions-sources/model-settings.png" alt="Model settings" />

6. Enter the **Description** and custom SQL **Query**. Click on **Run Query** to **Preview** the result.

<img src="../assets/warehouse-actions-sources/sample-sql-query.png" alt="Sample SQL query" />

<div class="infoBlock">

The <strong>Preview</strong> section displays the preview of the 50 resultant rows in a paginated format.
</div>

7. Click **Next** to configure the model succesfully.

## Update model

Follow these steps to update an existing model:

1. Click the model you want to update and select the **Configuration** tab.
2. Click on **Edit Configuration** and make the required changes in  **Description** and/or **Query**.
3. If you have edited the **Query**, click on **Run Query** to see if the new query generates the desired result.
4. Finally, click on **Finish** to save your changes.

<img src="../assets/warehouse-actions-sources/edit-model-configuration.png" alt="Edit existing Model configuration" />

<div class="infoBlock">
  

> If you update a model's <strong>Query</strong>, you also need to check or update the <a href="https://www.rudderstack.com/docs/warehouse-actions/visual-data-mapper/">Visual Data Mapper</a> mapping to trigger the full sync and validate a sync by manually triggering it.

Reviewer 2: This is not specific to Visual Data Mapper.
its true for both JSON and Visual Data Mapper.

</div>

## Model details

To get details about an existing model, you can click on it to see the options listed below:

- **Rename**: Lets you rename the model by clicking on the edit icon next to the model's name.

<img src="../assets/warehouse-actions-sources/model-rename.png" alt="Rename model" />

> - **Sources**: Displays the sources associated with the model. To know more on how to use a model with a source, refer to [Custom SQL Models](https://www.rudderstack.com/docs/warehouse-actions/custom-sql-models/).

Reviewer 2: The link and title has been changed. Also we can update the file name from `custom-sql-models.mdx`
